File: packages/imanikurd-quran/src/audio.ts
import type { Qari } from "./types.js";

/** Available Quran reciters with audio URL generators */
export const QARI_OPTIONS: Qari[] = [
  {
    id: "peshawa",
    name: "پێشەوا قادر (کوردی)",
    nameAr: "Peshawa Qadir",
    getUrl: (surahNum: number) => {
      const padded = surahNum.toString().padStart(3, "0");
      return `https://media.githubusercontent.com/media/w-coding/Peshawa-Qadir/main/mp3%20files/${padded}.mp3`;
    },
  },
  {
    id: "alafasy",
    name: "مشاری العفاسی",
    nameAr: "Mishary Alafasy",
    getUrl: (surahNum: number) => {
      const padded = surahNum.toString().padStart(3, "0");
      return `https://server8.mp3quran.net/afs/${padded}.mp3`;
    },
  },
  {
    id: "husary",
    name: "محمود خلیل الحصری",
    nameAr: "Mahmoud Al-Husary",
    getUrl: (surahNum: number) => {
      const padded = surahNum.toString().padStart(3, "0");
      return `https://server7.mp3quran.net/husr/${padded}.mp3`;
    },
  },
  {
    id: "minshawi",
    name: "محمد صدیق المنشاوی",
    nameAr: "Al-Minshawi",
    getUrl: (surahNum: number) => {
      const padded = surahNum.toString().padStart(3, "0");
      return `https://server10.mp3quran.net/minsh/${padded}.mp3`;
    },
  },
  {
    id: "abdulbasit",
    name: "عبدالباسط عبدالصمد",
    nameAr: "Abdul Basit",
    getUrl: (surahNum: number) => {
      const padded = surahNum.toString().padStart(3, "0");
      return `https://server7.mp3quran.net/basit/${padded}.mp3`;
    },
  },
  {
    id: "sudais",
    name: "عبدالرحمن السدیس",
    nameAr: "Abdulrahman Al-Sudais",
    getUrl: (surahNum: number) => {
      const padded = surahNum.toString().padStart(3, "0");
      return `https://server11.mp3quran.net/sds/${padded}.mp3`;
    },
  },
  {
    id: "shuraim",
    name: "سعود الشریم",
    nameAr: "Saud Al-Shuraim",
    getUrl: (surahNum: number) => {
      const padded = surahNum.toString().padStart(3, "0");
      return `https://server7.mp3quran.net/shur/${padded}.mp3`;
    },
  },
  {
    id: "maher",
    name: "ماهر المعیقلی",
    nameAr: "Maher Al-Muaiqly",
    getUrl: (surahNum: number) => {
      const padded = surahNum.toString().padStart(3, "0");
      return `https://server12.mp3quran.net/maher/${padded}.mp3`;
    },
  },
  {
    id: "ghamdi",
    name: "سعد الغامدی",
    nameAr: "Saad Al-Ghamdi",
    getUrl: (surahNum: number) => {
      const padded = surahNum.toString().padStart(3, "0");
      return `https://server7.mp3quran.net/s_gmd/${padded}.mp3`;
    },
  },
  {
    id: "ajamy",
    name: "أحمد العجمی",
    nameAr: "Ahmed Al-Ajmi",
    getUrl: (surahNum: number) => {
      const padded = surahNum.toString().padStart(3, "0");
      return `https://server10.mp3quran.net/ajm/${padded}.mp3`;
    },
  },
];

/** Get all available reciters */
export function getQariOptions(): Qari[] {
  return QARI_OPTIONS;
}

/** Get recitation URL for a surah by qari id */
export function getRecitationUrl(qariId: string, surahNumber: number): string | undefined {
  const qari = QARI_OPTIONS.find((q) => q.id === qariId);
  return qari?.getUrl(surahNumber);
}

/** Get a reciter by id */
export function getQari(qariId: string): Qari | undefined {
  return QARI_OPTIONS.find((q) => q.id === qariId);
}

export type { Qari };

File: packages/imanikurd_flutter/lib/src/audio.dart
import 'models.dart';

String _paddedSurah(int surahNum) => surahNum.toString().padLeft(3, '0');

String _peshawaUrl(int surahNum) =>
    'https://media.githubusercontent.com/media/w-coding/Peshawa-Qadir/main/mp3%20files/${_paddedSurah(surahNum)}.mp3';

String _alafasyUrl(int surahNum) => 'https://server8.mp3quran.net/afs/${_paddedSurah(surahNum)}.mp3';
String _husaryUrl(int surahNum) => 'https://server7.mp3quran.net/husr/${_paddedSurah(surahNum)}.mp3';
String _minshawiUrl(int surahNum) => 'https://server10.mp3quran.net/minsh/${_paddedSurah(surahNum)}.mp3';
String _abdulbasitUrl(int surahNum) => 'https://server7.mp3quran.net/basit/${_paddedSurah(surahNum)}.mp3';
String _sudaisUrl(int surahNum) => 'https://server11.mp3quran.net/sds/${_paddedSurah(surahNum)}.mp3';
String _shuraimUrl(int surahNum) => 'https://server7.mp3quran.net/shur/${_paddedSurah(surahNum)}.mp3';
String _maherUrl(int surahNum) => 'https://server12.mp3quran.net/maher/${_paddedSurah(surahNum)}.mp3';
String _ghamdiUrl(int surahNum) => 'https://server7.mp3quran.net/s_gmd/${_paddedSurah(surahNum)}.mp3';
String _ajamyUrl(int surahNum) => 'https://server10.mp3quran.net/ajm/${_paddedSurah(surahNum)}.mp3';

final List<Qari> qariOptions = [
  Qari(
    id: 'peshawa',
    name: 'پێشەوا قادر (کوردی)',
    nameAr: 'Peshawa Qadir',
    getUrl: _peshawaUrl,
  ),
  Qari(
    id: 'alafasy',
    name: 'مشاری العفاسی',
    nameAr: 'Mishary Alafasy',
    getUrl: _alafasyUrl,
  ),
  Qari(
    id: 'husary',
    name: 'محمود خلیل الحصری',
    nameAr: 'Mahmoud Al-Husary',
    getUrl: _husaryUrl,
  ),
  Qari(
    id: 'minshawi',
    name: 'محمد صدیق المنشاوی',
    nameAr: 'Al-Minshawi',
    getUrl: _minshawiUrl,
  ),
  Qari(
    id: 'abdulbasit',
    name: 'عبدالباسط عبدالصمد',
    nameAr: 'Abdul Basit',
    getUrl: _abdulbasitUrl,
  ),
  Qari(
    id: 'sudais',
    name: 'عبدالرحمن السدیس',
    nameAr: 'Abdulrahman Al-Sudais',
    getUrl: _sudaisUrl,
  ),
  Qari(
    id: 'shuraim',
    name: 'سعود الشریم',
    nameAr: 'Saud Al-Shuraim',
    getUrl: _shuraimUrl,
  ),
  Qari(
    id: 'maher',
    name: 'ماهر المعیقلی',
    nameAr: 'Maher Al-Muaiqly',
    getUrl: _maherUrl,
  ),
  Qari(
    id: 'ghamdi',
    name: 'سعد الغامدی',
    nameAr: 'Saad Al-Ghamdi',
    getUrl: _ghamdiUrl,
  ),
  Qari(
    id: 'ajamy',
    name: 'أحمد العجمی',
    nameAr: 'Ahmed Al-Ajmi',
    getUrl: _ajamyUrl,
  ),
];

List<Qari> getQariOptions() => qariOptions;

String? getRecitationUrl(String qariId, int surahNumber) {
  final qari = getQari(qariId);
  return qari?.getUrl(surahNumber);
}

Qari? getQari(String qariId) {
  for (final qari in qariOptions) {
    if (qari.id == qariId) return qari;
  }
  return null;
}
